reaper.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+func forkExec() {
+
+	// run supercronic in other process
+	pwd, err := os.Getwd()
+	if err != nil {
+		logrus.Fatalf("Failed to get current working directory: %s", err.Error())
+		return
+	}
+
+	pattrs := &syscall.ProcAttr{
+		Dir: pwd,
+		Env: os.Environ(),
+		Files: []uintptr{
+			uintptr(syscall.Stdin),
+			uintptr(syscall.Stdout),
+			uintptr(syscall.Stderr),
+		},
+	}
+
+	pid, err := syscall.ForkExec(os.Args[0], os.Args, pattrs)
+	if err != nil {
+		logrus.Fatalf("Failed to fork exec: %s", err.Error())
+		return
+	}
+
+	// forward signal to supercronic
+	signalToFork(pid)
+	// got supercronic exit status
+	wstatus := reapChildren(pid)
+	os.Exit(wstatus.ExitStatus())
+}
+
+func signalToFork(pid int) {
+	p, err := os.FindProcess(pid)
+	if err != nil {
+		logrus.Fatalf("Failed findProcess supercronic pid:%d,%s", pid, err.Error())
+	}
+	termChan := make(chan os.Signal, 1)
+	signal.Notify(termChan, siginalList...)
+	go func() {
+		for {
+			s := <-termChan
+			if err := p.Signal(s); err != nil {
+				logrus.Errorf("Failed to send signal to supercronic: %s", err.Error())
+			}
+		}
+	}()
+}
+
+// copy from https://github.com/ramr/go-reaper
+// modify for wait exit status of supercronic
+// without modify, supercronic exit status may not be obtained
+
+// Be a good parent - clean up behind the children.
+func reapChildren(superCrondPid int) syscall.WaitStatus {
+	var notifications = make(chan os.Signal, 1)
+
+	go sigChildHandler(notifications)
+
+	// all child
+	const pid = -1
+	var wstatus syscall.WaitStatus
+
+	for {
+		var sig = <-notifications
+		logrus.Debugf("reaper received signal %v\n", sig)
+		for {
+			pid, err := syscall.Wait4(pid, &wstatus, 0, nil)
+			for syscall.EINTR == err {
+				pid, err = syscall.Wait4(pid, &wstatus, 0, nil)
+			}
+
+			if syscall.ECHILD == err {
+				break
+			}
+
+			// when supercronic exit, may left some process
+			// expamle like "supercronic" -> B -> C
+			//            B exit, C may exist
```

Comment:
Is this what you're getting at?

```suggestion
			// when supercronic exits, it may leave some processes behind
			// this is OK since the reaper is running as pid 1 and will
			// terminate all processes when it exits
```

Reply:
I mean

> when Supercronic exits, it may leave some processes behind. This is okay since the reaper runs as a sidecar for Supercronic.

https://github.com/aptible/supercronic/blob/master/cron/cron.go#L72

Reply:
Right, I'm trying to determine why that's supposedly okay. The reaper is programmed to exit as soon as supercronic does. It won't clean up any remaining processes. The only way I see that being okay is if it's pid 1 which means when it exits, in a container, it cleans up the remaining processes. Otherwise it wouldn't be okay, correct?

Reply:
Sorry, I'm not saying the code needs to change, the original comment had me concerned because it was talking about processes being left behind which seemed like something that the reaper process should be handling. I now understand that's not what the reaper does, it cleans up after processes as they exit.

However, I do find the current comment to be misleading:

> when Supercronic exits, it may leave some processes behind. This is okay since the reaper runs as a sidecar for Supercronic.

To me, this implies that we don't have to worry about superconic leaving processes behind because the reaper will still be running but the reaper exits too. I don't want someone else to see this and think that the code is wrong because we do want the reaper to exit when superconic does.

We don't need to mention the fact that processes may be left behind because they'll be cleaned up when the reaper, as pid 1, exits. Just something simple like `// Have the reaper exit when supercronic does` or no comment is sufficient.

Reply:
Ok, i remove the comment.